text/readable/readable.1
.TH readable 1 "readable" "27 Oct 2016" "v1 2016.10.27"

.SH NAME
readable - heuristic garbage string filter

.SH SYNOPSIS
command | \fBreadable\fP [\fB-c\fP CUTOFF]

.SH DESCRIPTION
\fBreadable\fP reads standard input line by line, computes "readability index"
for each one of them, normalizes it to the unit interval [\fB0\fP, \fB1\fP],
discards lines below a certain readability index, and prints remaining ones.

.SH ALGORITHM
This version of \fBreadable\fP implements a rather naive readability index
algorithm. It only looks at kinds of characters present, favoring letters and
disliking other symbols, and does not take into account their order or
grouping.

.SH OPTIONS
.TP
\fB\-c\fP CUTOFF, \fB\-\-cutoff\fP=CUTOFF
Set the cutoff readability index. All lines with index less than \fBCUTOFF\fP
will be discarded. This must be a floating-point number between 0 and 1.

Default: \fB0.5\fP.

.TP
\fB\-h\fP, \fB\-\-help\fP
Print a short help message.

.SH AUTHOR
Developed by Ilya Terentyev <bacondropped@gmail.com> as a part of the
\fBbacontools\fP package <https://github.com/bacondropped/bacontools>.

.SH LICENSE
\fBreadable\fP is distributed under the terms of the MIT license.

.SH SEE ALSO
\fBbacontools\fP(7)


CHANGELOG.md
bacontools change log
=====================
Notable changes to `bacontools` are documented in this file.

***

[2016.10.27]
------------
### Added
+ `togif`: implement $NO_OVERWRITE
+ `stopwatch`: print total laps on SIGINT
+ `stopwatch`: laps
+ `git-force-pristine`
+ `xclip-tofile`
+ `tasktags`: use less as pager
+ `git-ls`: `--noalign` option

### Changed
+ `togif`: prettier notifications
+ `miniscripts/batch`: $DIRECTORY -> $1
+ `imgur-dl`: use custom User-Agent
+ `cate`: pluralize line count properly
+ `wget-parallel`: rename *_OPTIONS to *_ARGS
+ `wget-parallel`: also read URLs from stdin
+ `wget-parallel`: sort URLs and discard duplicates

### Fixed
+ `devbup`: correctly set dcfldd as the default dd
+ `myzuka-dl`: catch EOFError (Ctrl-D in prompts)
+ `wget-parallel`: fix $#=0 case
+ `wget-parallel`: fix newline separation
+ `imgur-dl`: trap SIGINT
+ `git-ls`: discard GPG verification in `git log` output

***

[2016.09.27]
------------
### Added
+ `miniscripts/git-ls-sorted`
+ `checktor`
+ `content-length`
+ `zerofile`
+ `cursorwiggle`
+ `imgur-dl`: -u/--unique option
+ `bindump`
+ `imgur-dl`: read `$TIMEOUT` from environment
+ `clone-github-user`: read `$PER_PAGE` from environment
+ `clone-github-user`: pass arguments to `git clone`
+ `wget-parallel`
+ `devbup`: implement `$PADDING_SOURCE`
+ `preptrack`: warn if ID3v1 tags are too long
+ `cate`

### Changed
+ `git-ls`: separate columns by tab instead of single space
+ `Makefile`: don't run `patches/apply` in all
+ extract Git-specific utilities to `git/`
+ `imgur-dl`: smarter messages
+ `vimless`: check if input files are readable

### Fixed
+ `preptrack`: default to empty tags if exceptions are thrown
+ `imgur-dl`: strip path slashes properly
+ `baconplaylist`: remove hardcoded path
+ `devbup`: fix '$CKSUM_COMMAND is ""' error
+ `imgur-dl`: don't create a subfolder when no images are extracted
+ `imgur-dl`: refactor image extraction (now finally works with GIF posts)
+ `imgur-dl`: fix exceptions on single image posts
+ `togif`: fix incorrect quoting in requisite error report

***

[2016.08.26]
------------
### Added
+ `devbup`: check if *_COMMAND commands exist
+ `watchclock`
+ `patches/check`
+ `apdiff.1` manpage
+ `miniscripts/`
+ `keepalive`
+ `snippets/vagrant.md`: Vagrantfile snippets
+ `preptrack`: `V1` environment variable
+ `docker-cleanup`
+ `bacontools.7`: add NOTES section
+ `preptrack`: check dependencies

### Changed
+ `miniscripts/batch`: don't put dots after "e" for episodes

### Fixed
+ `preptrack`: escape single quotes in filenames
+ `clone-github-user`: ensure unique URLs
+ `preptrack`: fix Python exception when converting videos
+ `bananaglee`: install from Makefile (fixes install on Arch)
+ `preptrack`: use inline Python code for latest eyeD3

***

[2016.08.17]
------------
### Added
+ `apdiff`: `GLOB` environment variable
+ `bacontools.7` manpage
+ `devbup`: implement `DONT_SHRED`
+ `git-stat-atr`
+ `center`

### Changed
+ `update-all`: use `figlet` instead of `toilet`
+ `devbup`: don't split when `SPLIT_SIZE` = `0`
+ `PKGBUILD`: install `dev/main` version

### Fixed
+ `devbup`: run checksum under sudo to avoid permission problems
+ `devrestore`: fix GPG decryption with anon recipients

***

[2016.07.27]
------------
### Added
+ `clone-github-user`
+ `httpdf`: `-path` command line argument
+ `cptemp`
+ `PKGBUILD`
+ `myzuka-dl`: `MAX_RESULTS` and `TIMEOUT` environment variables

### Fixed
+ `single-urxvt`: minimize properly on KDE (requires `xdotool`)
+ `imgur-dl`: specify correct manpage directory
+ `devbup`: support non-root execution
+ `git-ls`: specify Python 2 in shebang (helps when default Python is v3)
+ `myzuka-dl`: don't exit when unable to download one track

***

[2016.07.15]
------------
### Added
+ `togif`: manpage
+ `git-repo-list`
+ `wget-page`
+ `stopwatch`
+ `readable`: manpage
+ `vimless`: `-N` option for less; line numbers are now not printed by default

### Changed
+ `checkreboot`: don't print "reboot-required.pkgs" prompt
+ `bitcount`: display true bits / all bits ratio
+ `vimless`: loop over each file argument
+ `baconplaylist`: ignore case in artist names
+ `baconplaylist`: display top N artists
+ `git-ls`: separate cols with 1 space instead of 2
+ `vimless`: exit if $# = 0 and stdin is a tty
+ `myzuka`: print output except filenames to stderr
+ `togif`: use `set -e` to check for failing commands

### Fixed
+ `git-ls`: verify that argument paths exist
+ `git-ls`: list subdirectories and files correctly

***

[2016.07.05]
------------
### Added
+ `curl-tt`
+ `bincounter`: total bytes and normalized bit counts
+ `morseconv`
+ `baconplaylist`
+ `vimless`
+ `CHECKLIST.md`
+ `togif`: implement `$QUIET` and `$NOSTATS`
+ `snippets/shell.md`

### Changed
+ `bincounter` -> `bitcount`

***

[2016.06.29]
------------
### Added
+ `single-urxvt`

### Changed
+ `netinfo`: deduce active interface via `ip link`
+ `devbup`: prefer `dcfldd` if it's installed

### Fixed
+ `netinfo`: check if nmcli is installed
+ `preptrack`: check if files exist

***

[2016.06.17]
------------
### Added
+ `myzuka-dl`: handle `DIR` environment variable
+ `du1`
+ `respcode`
+ `ringcat`
+ `bincounter`
+ `preptrack`: notify-send notification
+ `maturity.txt`
+ `devrestore`

### Changed
+ Makefile: run `patches/apply` on install
+ `devbup`: move up from Untested to Moving

### Fixed
+ `togif`, `preptrack`: fix `$(command ...)` invocations
+ `imgur-dl`: downgrade to commit dfd26dc
+ `devbup`: fix no shredding bug
+ `imgur-dl`, `myzuka-dl`: eschew Bundler on Raspbian with a patch
+ `togif`: check if dependencies are installed
+ `devbup`: run MAT on each file before tarballing

***

[2016.05]
---------
### Added
+ `httpdf`
+ `TOOLS.md`
+ `preptrack`: handle embedded pictures
+ `myzuka-dl`: `-n`/`--index` option
+ `myzuka-dl`: Gemfile
+ `imgur-dl`: Gemfile
+ `preptrack`: man page
+ Tool maturity classification
+ `git-ls`: support arbitrary paths
+ `patches` directory, `no-haskell` patch
+ `userls`
+ `myzuka`: `-d`/`--dir` option

text/readable/readable.c
#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>

#define READABLE_VERSION "1"

const char *short_options = "hc:";

struct option long_options[] = {
	{"help",   no_argument,       0, 'h'},
	{"cutoff", required_argument, 0, 'c'},
	{0,        0,                 0,  0 }
};

const char *help_text = ""\
"bacontools readable v" READABLE_VERSION "\n"
"Usage: strings some_file | readable [options...]\n"\
"Filter out unreadable (garbage) lines.\n\n"\
"COMMAND LINE OPTIONS:\n"\
" -h, --help                   Print this help text.\n"\
" -c, --cutoff                 Set the minimum readability value (default 0.5).\n"\
"";

double compute_r_index(
	const char *string
);

double compute_r_index(const char *string) {
	double result = 0;

	unsigned int lowcases = 0;
	unsigned int upcases  = 0;
	unsigned int digits   = 0;
	unsigned int others   = 0;

	unsigned int i = 0;
	unsigned int total = (unsigned int)strlen(string);

	for (i = 0; i < total - 1; i++) {
		char cur = string[i];

		// TODO: Smarter algorithm, with sequential vowels/consonants,
		// mixing of upper and lower case, spacing, etc.
		if (cur >= 'a' && cur <= 'z') {lowcases++;}
		else if (cur >= 'A' && cur <= 'Z') {upcases++;}
		else if (cur >= '0' && cur <= '9') {digits++;}
		else others++;
	};

	// BUG: Doesn't produce exact results, maybe due to IEEE754 error
	// accumulation, or maybe I just fucked this one up
	result = (2.0*lowcases
	          + 1*upcases
	        - 0.8*digits
	          - 2*others)
	/total/4+0.5;

	return result;
};

int main(int argc, char **argv) {
	char *line = NULL;
	size_t size;
	int c = 0;

	double cutoff = 0.5;

	while(1) {
		int opt_index = 0;
		c = getopt_long(argc, argv, short_options, long_options, &opt_index);
		if (c == -1) break;

		switch(c) {
		case 'h':
			printf("%s", help_text);
			exit(EXIT_SUCCESS);
		case 'c':
			cutoff = strtod(optarg, NULL);
			break;
		};
	};

	if (optind < argc) {
		fprintf(stderr, "readable: unexpected non-option "\
		                "argument %s\n", argv[optind]);
		exit(EXIT_FAILURE);
	};

	while (getline(&line, &size, stdin) != -1) {
		double r_index = compute_r_index(line);
		if (r_index >= cutoff) printf("%s", line);
	};

	return EXIT_SUCCESS;
};
